src/NuGet.Core/NuGet.DependencyResolver.Core/GraphModel/Tracker.cs

```diff
@@ -3,83 +3,136 @@
         }
 
         public void MarkAmbiguous(GraphItem<TItem> item)
         {
-            GetEntry(item).Ambiguous = true;
+            GetOrAddEntry(item).Ambiguous = true;
         }
 
         public bool IsBestVersion(GraphItem<TItem> item)
         {
-            var entry = GetEntry(item);
-
-            var version = item.Key.Version;
+            var entry = TryGetEntry(item);
 
-            foreach (var known in entry.List)
+            if (entry is not null)
             {
-                if (version < known.Key.Version)
+                var version = item.Key.Version;
+
+                foreach (var known in entry.Items)
                 {
-                    return false;
+                    if (version < known.Key.Version)
+                    {
+                        return false;
+                    }
                 }
             }
 
             return true;
         }
 
-        public IEnumerable<GraphItem<TItem>> GetDisputes(GraphItem<TItem> item) => GetEntry(item).List;
+        public IEnumerable<GraphItem<TItem>> GetDisputes(GraphItem<TItem> item)
+        {
+            return TryGetEntry(item)?.Items ?? Enumerable.Empty<GraphItem<TItem>>();
+        }
 
         internal void Clear()
         {
-            _entries.Clear();
+            _entryByLibraryName.Clear();
         }
 
-        private Entry GetEntry(GraphItem<TItem> item)
+        private Entry? TryGetEntry(GraphItem<TItem> item)
         {
-            Entry itemList;
-            if (!_entries.TryGetValue(item.Key.Name, out itemList))
+            _entryByLibraryName.TryGetValue(item.Key.Name, out Entry? entry);
+            return entry;
+        }
+
+        private Entry GetOrAddEntry(GraphItem<TItem> item)
+        {
+            if (!_entryByLibraryName.TryGetValue(item.Key.Name, out Entry? entry))
             {
-                itemList = new Entry();
-                _entries[item.Key.Name] = itemList;
+                entry = new Entry();
+                _entryByLibraryName[item.Key.Name] = entry;
             }
-            return itemList;
+
+            return entry;
         }
 
-        private class Entry
+        private sealed class Entry
         {
-            public Entry()
+            /// <summary>
+            /// This field can have one of three values:
+            ///
+            /// - null: when no graph items exist in this entry
+            /// - a graph item: when only a single graph item exists in this entry
+            /// - a hash set of graph items: when multiple graph items exist in this entry
+            ///
+            /// This packing exists in order to reduce the size of this object in
+            /// memory. For large graphs, this can amount to a significant saving in memory,
+            /// which helps overall performance.
+            /// </summary>
+            private object? _storage;
+
+            public bool Ambiguous { get; set; }
+
+            public void AddItem(GraphItem<TItem> item)
```

Comment:
I know we're testing it indirectly, but wouldn't it be easier to get some times for this class explicitly and this particular method? 

It's where the major change is.

Reply:
I feel like it's thoroughly tested via the API of `Tracker`. This class is really an internal implementation detail of that API, similar to [`Dictionary<,>.Entry`](https://github.com/microsoft/referencesource/blob/51cf7850defa8a17d815b4700b67116e3fa283c2/mscorlib/system/collections/generic/dictionary.cs#L61-L66).